### DISCUSSION TITLE
Plugins!

### DISCUSSION THREAD

**Discussion by inorganik** [inorganik/countUp.js#306] [Ideas]
Recently, @msoler75 made PR #304 into CountUp with a new animation style:

![marcel-countup](https://user-images.githubusercontent.com/109553/224044898-c101b360-f4ba-4da2-ab66-de2c9c850736.gif)

It's very cool! But rather than merge his code, (about 130 lines) I would like to create a plugin interface for CountUp. This would allow plugin authors to get credit for and maintain their code, prevent shipping unused code, and allow for endless alternative animations of CountUp.

As part of supporting plugins, I would like to call out each available plugin in the README of CountUp with a sample gif, explanation, package name and author credit. I would also add an explanation of how to use plugins in general.

Which brings me to the API. Marcel's code gets invoked in `printValue`, and just needs a reference to `this.el`. So if plugins where exported as a class with a render function, we could use it like this:

```ts
import { CountUp } from 'countup.js';
import { Odometer } from 'countup-odometer'; // just an idea for the name

const odometerPlugin = new Odometer({
  duration: 1,
  delay: 0.3,
  // etc... any options the plugin author wants
});
const numAnim = new CountUp('targetId', 5234, {
  plugin: odometerPlugin
});
numAnim.start();
```

The plugin interface could look like this. It just needs to implement a render function that accepts the target element and a string of the formatted number. Render would get called in place of the existing `printValue` on each frame.
```ts
export declare interface CountUpPlugin {
   render(elem: HTMLElement | HTMLInputElement, formatted: string): void;
}
```

So the Odometer class might look something like this:
```ts
export class Odometer implements CountUpPlugin {
  // ...some properties here

  constructor(options: OdometerOptions) {
    // ...setup code
  }

  render(elem: HTMLElement | HTMLInputElement, formatted: string): void {
    // render DOM here
  }
}
```

Thoughts?

**Comment by inorganik** [comment]
It occurred to me that this particular plugin would not support input elements... that's ok.

**Comment by msoler75** [comment]
Good suggestion. I am waiting for the next version to create the odometer plugin with this simple and effective interface :)

**Comment by msoler75** [comment]
Ready! You can check https://github.com/msoler75/odometer_countup.js and test it?

**Reply by inorganik** [reply]
This looks good. Were you able to publish an npm package?

**Comment by inorganik** [comment]
Looks good to me! Can you publish a package on npm and add install instructions to your repo? Once that is done I will update the README and link to your repo.

**Comment by inorganik** [comment]
@msoler75 just wanted to check in. Were you able to publish your plugin? I'd love to add it to the countUp README

**Reply by msoler75** [reply]
Hello, sorry for the delay, here is the package [odometer_countup](https://www.npmjs.com/package/odometer_countup)

**Reply by inorganik** [reply]
Awesome! Thanks Marcel. I will update the README on CountUp. One small feedback, I see in your example you are importing the packages from urls, but in most every use-case they will import from the package, like the following:
```ts
import { Odometer } from 'odometer_countup';
```

**Reply by inorganik** [reply]
Also, looks like the default distribution is umd, so you are not able to import it like that anyway...   I will try to make a PR soon into your repo to address those things
